families/track_and_trade/sawtooth_track_and_trade/handler.py

```diff
@@ -0,0 +1,633 @@
+        container = AgentContainer()
+        container.entries.extend([agent])
+
+        _set_container(state, address, container)
+
+        return
+
+    # A container exists at the agent's address. Check to see if the
+    # agent has already been created.
+    for agent in container.entries:
+        if agent.public_key == signer:
+            raise InvalidTransaction(
+                'Agent already exists with key {}...'.format(
+                    signer[:6]))
+
+    # The agent doesn't exist. Create it and put it in the container.
+    LOGGER.debug(
+        'Creating agent -- key: {} -- name: {}'.format(
+            signer[:6], name))
+
+    agent = Agent(
+        public_key=signer,
+        name=name,
+        timestamp=timestamp,
+    )
+
+    container.entries.extend([agent])
+    container.entries.sort(key=lambda ag: ag.public_key)
+
+    _set_container(state, address, container)
+
+
+def _create_record(payload, signer, timestamp, state):
+    '''
+    * Check that the signer is registered.
+    * Check that the record doesn't already exist.
+    * Check that the record type exists.
+    * Check that the required properties have been provided.
+    * Create the record (and record container if needed).
+    * Create the associated properties.
+    * Post the provided property values.
+    '''
+
+    # Check that the signer is registered.
+    _verify_agent(state, signer)
+
+    txn = payload.create_record
+
+    # Check that the record doesn't already exist
+    record_id = txn.record_id
+
+    record_address = Addressing.make_record_address(record_id)
+    record_container = _get_container(state, record_address)
+
+    if record_container is not None:
+        if any([record.record_id == record_id
+                for record in record_container.entries]):
+            raise InvalidTransaction(
+                'Record {} already exists'.format(record_id))
+
+    # Check that the record type exists.
+    record_type = txn.record_type
+    type_address = Addressing.make_record_type_address(record_type)
+    type_container = _get_container(state, type_address)
+
+    if type_container is None or all([
+            rec_type.name != record_type
+            for rec_type in type_container.entries]):
+        raise InvalidTransaction(
+            'No record type {} exists'.format(record_type))
+
+    # Record type exists -- check the required properties
+    for rec_type in type_container.entries:
+        if rec_type.name == record_type:
+            type_properties = rec_type.properties
+            break
+
+    type_schemata = {
+        prop.name: prop
+        for prop in type_properties
+    }
+
+    required_properties = [
+        prop
+        for prop in type_properties
+        if prop.required
+    ]
+
+    provided_properties = {
+        prop.name: prop
+        for prop in txn.properties
+    }
+
+    # Make sure the required properties are all provided
+    required_names = [prop.name for prop in required_properties]
```

Comment:
I am actually not a fan of creating a bunch of variables like this when they are only used once, and the value they represent (a fairly simple list comprehension) is pretty straightforward.

The problem is you have dozens of different variables in this function, all with very similar names, and I am losing track of which is what. To me, `[prop.name for prop in required_properties]` has more meaning than `required_names`.

I would look at which of these variables could be removed to make this function more readable.